### ISSUE TITLE
Python script doesn't progress once ad appears on screen

### ISSUE THREAD

**Issue by robert-altom** [robert-altom/test#831]
In order for my test script to be successful, it has to navigate past an interstitial ad loaded by the MoPub Unity plugin. However, once the ad is shown on screen, there seems to be some sort of hang up. The app closes on the device (back to the home screen) and on the Python console, nothing progresses from the previous step. It just hangs. Then, I have to manually cancel the script. Yet, I get no error message.

---
<sub>You can find the original issue from GitLab [here](https://gitlab.com/altom/altunity/altunitytester/-/issues/58).</sub>

**Comment by robert-altom** [comment]
AltUnity Tester cannot interact with native elements. Assuming that this is the case here with mopub. 
The workaround here is to use appium to interact with the native elements. Then bring the application back in focus. Once its back in focus you can continue to interact with the game using AltUnity Tester.

**Comment by robert-altom** [comment]
So, what if `OnApplicationPause` was used to create a temporary band-aid connection for AltUnityTester? Not being overly familiar with your plugin code, I am not sure exactly how that would look. Fundamentally, however, a method could be created in the games GameController script to programmatically create an AltUnityTester connection if `OnApplicationPause` returns true.

Actually, the AltUnityTester connection would probably have to be created as a fragment of the ad plugin. Then, it would be destroyed when the ad is dismissed...

**Comment by robert-altom** [comment]
My co-developer and I are looking at a possible workaround using `OnApplicationPause`. I will post here once we figure something out.

**Comment by robert-altom** [comment]
@jas85 We have thought about `OnApplicationPause`, but I think all we can do from AltUnityTester is to wait until the app is no longer paused. So the ad would still need to be closed by some other means. Or is there maybe a "timeout" that occurs once the ad is played? That could be useful. 

We will look into `OnApplicationPause` in more detail to see if there's anything more we can do (other than wait).

**Comment by robert-altom** [comment]
The workaround should work because the command that are send to appium are not going through the prefab of AltUnityTester that is paused. 

Maybe I didn't understand your last line correctly so maybe my explanation will be completely useless but when I when was talking about Unity Player I am not saying Unity Editor but the activity `com.unity3d.player.UnityPlayerActivity` the Unity create when making the apk.

**Comment by robert-altom** [comment]
And the more I think of it, the issue as you've described it would not need a solution to be implemented into AltUnityTester, it would need a solution to be implemented into the game code itself. 

Unity offers a method for such an issue: https://answers.unity.com/questions/1289539/prevent-android-pause-when-an-overlay-is-active.html

Now, the issue is figuring out how to use that method to keep AltUnityTester active while the ad is on the screen.

**Comment by robert-altom** [comment]
Okay, I understand your explanation for why this is happening. However, your suggested workaround seems to contradict that explanation. If AltUnityTester is paused when an ad appears, how is that statement going to make any difference...when it will need AltUnityTester to execute it?

Also, I am not sure how you've concluded this issue is Unity Player specific. I am utilizing the Python bindings to execute an external test script.

**Comment by robert-altom** [comment]
Hi, I come back with some update:

When the interstitial ad loaded by the MoPub Unity plugin the activity manager changes the activity from UnityPlayer to the ad. Because of that the game is paused and with that also AltUnityTester. For now we are interested for AltUnityTester to be able to test only in UnityPlayer, so we are not going to implement anything in this regard in the near future(maybe only a mechanism to signal that that the game is paused if the ads are loaded randomly).

A workaround for your problem is to use `self.altdriver.appium_driver.press_keycode(4)` when the ad appear. 4 is the key code for back button. This way it will return to Unity app and the test will proceed. 

I hope the workaround will work in your case.

**Comment by robert-altom** [comment]
Hi, thank you for the update. From what I've seen in my investigation it looks like AltUnityTester is no longer able to send back responses once the ad appears, but after you close manually the ad the test will resume. This is indeed unacceptable and I will be looking for a solution.

**Comment by robert-altom** [comment]
As an update, I discovered that the reason for the app closing on the device was because of Appium's command timeout parameter. By default, it is set to 60 seconds. So, I simply implemented this statement to increase it:

`self.desired_caps['newCommandTimeout'] = '90'`

Still, though, the issue remains that once an ad appears on screen, the Python script seems to stall out. Now, after 90 seconds, the same thing happens as I described above.
